### PR TITLE
Log data loading errors in conversion / mapping application

### LINKED ISSUE #8081: log all data loading errors
Bucket requests always return 200, if something went wrong during loading, the array is all-black and the MISSING-BUCKETS header is set. However, no error information is contained.

There is already good error logging in `BinaryDataService.handleBucketRequest` for everything that happens inside of `BucketProvider.load`.

However, if the errors occur after that, e.g. while applying the agglomerate mapping, nothing is shown. This error logging should be moved to `handleDataRequests` or another type of error logging should be created at that level there. Maybe include the errors in the response headers? (Or is that too verbose/can sensitive information be leaked?)

### PR DESCRIPTION
**Pull request by frcroth**
### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- Add an exception to the conversion

### Issues:
- fixes #8081 

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [x] Removed dev-only changes like prints and application.conf edits
- [ ] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [x] Needs datastore update after deployment

### CODE REVIEW

**Review by fm3 — COMMENTED** [review]
Nice, does the trick!
I wonder if we can make this a little dryer. Maybe the `datasetErrorLoggingService` could have a single public function, something like
`withErrorLogging[T](dataSourceId: DataSourceId, label: String, resultFox: Fox[T]): Fox[T]`. This could be called like
`datasetErrorLoggingService.withErrorLogging(request.dataSource.id, "Converting data", conversionFunc(inputArray))`. What do you think? (Note that I did not try this, so it may not work exactly in this way).

**Comment by frcroth** [comment]
> Nice, does the trick! I wonder if we can make this a little dryer. Maybe the `datasetErrorLoggingService` could have a single public function, something like `withErrorLogging[T](dataSourceId: DataSourceId, label: String, resultFox: Fox[T]): Fox[T]`. This could be called like `datasetErrorLoggingService.withErrorLogging(request.dataSource.id, "Converting data", conversionFunc(inputArray))`. What do you think? (Note that I did not try this, so it may not work exactly in this way).

I assume this would then be used for both conversion and handleBucketRequest?
One problem is that the applicationHealthService is not available there, so should this be passed in? Should there still be a matching for internalError then?
Additional notes:
- This needs to check for datasetErrorLoggingService exists (is an option)
- T needs to be Array[Byte] so that we can still check for zero length array
- convertIfNecessary needs to return Fox

I implemented it in such a way. What do you think?

**Comment by fm3** [comment]
Yes, that’s what I had in mind :)

Let’s remove applicationHealthService from the BinaryDataService and inject it in DatasetErrorLoggingService directly instead. That should make this a little more simple.

The other points are correct (if a little unfortunate). Let’s go with that :)

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/DatasetErrorLoggingService.scala [hunk 1]
```diff
@@ -41,4 +45,39 @@ class DatasetErrorLoggingService @Inject()(
...
+      case Failure(msg, Full(e: InternalError), _) =>
+        logger.error(s"Caught internal error: $label for $dataSourceId", e)
+        onInternalError(e)
+        Fox.failure(msg, Full(e))
+      case Failure(msg, Full(exception), _) =>
+        if (shouldLog(dataSourceId.team, dataSourceId.name)) {
+          logger.error(s"Error while $label for $dataSourceId Stack trace: ${TextUtils.stackTraceAsString(exception)} ")
+          registerLogged(dataSourceId.team, dataSourceId.name)
+        }
+        Fox.failure(msg, Full(exception))
```
[the remark this answers is not in the document]
Comment (frcroth) [reply webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/DatasetErrorLoggingService.scala:72]: I tried this.

Using built in exception printing:

```
backend: 2024-11-25 10:28:06,075 [error] com.scalableminds.webknossos.datastore.services.DatasetErrorLoggingService - Error while Bucket loading for layer N5-Upload at BucketPosition(voxelMag1 at (650, 26, 775), bucket at (20,0,24), mag(1, 1, 1)), cuboid: Cuboid((650, -38, 775) / (1, 1, 1),200,200,1) for DataSourceId(sample_organization/N5-Upload)
backend: java.lang.Exception: Absolute path /home/felix/scm/webknossos/binaryData/sample_organization/N5-Upload/s3 in local file system is not in path whitelist. Consider adding it to datastore.localFolderWhitelist
backend:        at com.scalableminds.webknossos.datastore.storage.RemoteSourceDescriptorService.$anonfun$uriForMagLocator$1(RemoteSourceDescriptorService.scala:68)
backend:        at net.liftweb.common.Tryo.tryo(Tryo.scala:22)
backend:        at net.liftweb.common.Tryo.tryo$(Tryo.scala:20)
backend:        at net.liftweb.common.Box$.tryo(Box.scala:56)
backend:        at net.liftweb.common.Tryo.tryo(Tryo.scala:64)
backend:        at net.liftweb.common.Tryo.tryo$(Tryo.scala:64)
backend:        at net.liftweb.common.Box$.tryo(Box.scala:56)
backend:        at com.scalableminds.webknossos.datastore.storage.RemoteSourceDescriptorService.uriForMagLocator(RemoteSourceDescriptorService.scala:52)
backend:        at com.scalableminds.webknossos.datastore.storage.RemoteSourceDescriptorService.$anonfun$remoteSourceDescriptorFor$1(RemoteSourceDescriptorService.scala:45)
backend:        at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:470)
```

With Fox TextUtils stackTrace

```
backend: 2024-11-25 10:57:23,254 [error] com.scalableminds.webknossos.datastore.services.DatasetErrorLoggingService - Error while Bucket loading for layer N5-Upload at BucketPosition(voxelMag1 at (1120, 32, 1152), bucket at (35,1,36), mag(1, 1, 1)), cuboid: Cuboid((1120, 32, 1152) / (1, 1, 1),32,32,32) for DataSourceId(sample_organization/N5-Upload) Stack trace: java.lang.Exception: Absolute path /home/felix/scm/webknossos/binaryData/sample_organization/N5-Upload/s3 in local file system is not in path whitelist. Consider adding it to datastore.localFolderWhitelist
backend:        at com.scalableminds.webknossos.datastore.storage.RemoteSourceDescriptorService.$anonfun$uriForMagLocator$1(RemoteSourceDescriptorService.scala:68)
backend:        at net.liftweb.common.Tryo.tryo(Tryo.scala:22)
backend:        at net.liftweb.common.Tryo.tryo$(Tryo.scala:20)
backend:        at net.liftweb.common.Box$.tryo(Box.scala:56)
backend:        at net.liftweb.common.Tryo.tryo(Tryo.scala:64)
backend:        at net.liftweb.common.Tryo.tryo$(Tryo.scala:64)
backend:        at net.liftweb.common.Box$.tryo(Box.scala:56)
backend:        at com.scalableminds.webknossos.datastore.storage.RemoteSourceDescriptorService.uriForMagLocator(RemoteSourceDescriptorService.scala:52)
backend:        at com.scalableminds.webknossos.datastore.storage.RemoteSourceDescriptorService.$anonfun$remoteSourceDescriptorFor$1(RemoteSourceDescriptorService.scala:45)
backend:        at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:470)
backend:        at org.apache.pekko.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:73)
backend:        at org.apache.pekko.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:110)
backend:        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
backend:        at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
backend:        at org.apache.pekko.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:110)
backend:        at org.apache.pekko.dispatch.TaskInvocation.run(AbstractDispatcher.scala:59)
backend:        at org.apache.pekko.dispatch.ForkJoinExecutorConfigurator$PekkoForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:57)
backend:        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
backend:        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
backend:        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
backend:        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
backend:        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
```

So the second thing is longer, but it is not clear if there is any useful information in the additional trace.

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/DatasetErrorLoggingService.scala [hunk 2]
```diff
@@ -41,4 +46,36 @@ class DatasetErrorLoggingService @Inject()(
...
+          if (shouldLog(dataSourceId.team, dataSourceId.name)) {
+            logger.warn(msg)
+            registerLogged(dataSourceId.team, dataSourceId.name)
+          }
+          Fox.failure(msg)
+        } else {
+          Fox.successful(data)
+        }
+      case Failure(msg, Full(e: InternalError), _) =>
+        logger.error(s"Caught internal error: $label for $dataSourceId", e)
```
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/DatasetErrorLoggingService.scala:64]:
```suggestion
        logger.error(s"Caught internal error while $label for $dataSourceId:", e)
```

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/BinaryDataService.scala [hunk 3]
```diff
@@ -91,28 +96,14 @@ class BinaryDataService(val dataBaseDir: Path,
...
-              s"Bucket provider returned Full, but data is zero-length array. Layer ${request.dataLayer.name} of dataset ${request.dataSource.id}, ${request.cuboid}"
-            logger.warn(msg)
-            Fox.failure(msg)
-          } else Fox.successful(data)
-        case other => other.toFox
+      datasetErrorLoggingService match {
+        case Some(d) =>
+          d.withErrorLogging(
+            request.dataSource.id,
+            s"Bucket loading for layer ${request.dataLayer.name} at ${readInstruction.bucket}, cuboid: ${request.cuboid}",
```
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/BinaryDataService.scala:103]:
```suggestion
            s"loading bucket for layer ${request.dataLayer.name} at ${readInstruction.bucket}, cuboid: ${request.cuboid}",
```

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/BinaryDataService.scala [hunk 4]
```diff
@@ -55,22 +54,28 @@ class BinaryDataService(val dataBaseDir: Path,
   def handleDataRequests(requests: List[DataServiceDataRequest]): Fox[(Array[Byte], List[Int])] = {
     def convertIfNecessary(isNecessary: Boolean,
                            inputArray: Array[Byte],
-                           conversionFunc: Array[Byte] => Box[Array[Byte]]): Box[Array[Byte]] =
-      if (isNecessary) conversionFunc(inputArray) else Full(inputArray)
+                           conversionFunc: Array[Byte] => Fox[Array[Byte]],
+                           request: DataServiceDataRequest): Fox[Array[Byte]] =
+      if (isNecessary) datasetErrorLoggingService match {
+        case Some(value) => value.withErrorLogging(request.dataSource.id, "Conversion", conversionFunc(inputArray))
```
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/BinaryDataService.scala:60]:
```suggestion
        case Some(value) => value.withErrorLogging(request.dataSource.id, "converting bucket data", conversionFunc(inputArray))
```

**Review by fm3 — APPROVED** [review]
Nice, works for me :) I just added some small suggestions for smoother grammar in the logging output.

**Pull request merged**